airbyte-integrations/connectors/source-intercom/source_intercom/source.py

```diff
@@ -0,0 +1,364 @@
+        data = response.json()
+
+        for data_field in self.data_fields:
+            if data and isinstance(data, dict):
+                data = data.get(data_field, [])
+
+        if isinstance(data, list):
+            data = data
+        elif isinstance(data, dict):
+            data = [data]
+
+        return data
+
+    def parse_response(
+        self,
+        response: requests.Response,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping]:
+
+        data = self.get_data(response)
+
+        for record in data:
+            yield record
+
+    def backoff_time(self, response: requests.Response) -> Optional[float]:
+        wait_time = 3600 / self.queries_per_hour  # wait for 3,6 seconds according to API limit
+        return wait_time
+
+
+class IncrementalIntercomStream(IntercomStream, ABC):
+    cursor_field = "updated_at"
+
+    def filter_by_state(self, stream_state: Mapping[str, Any] = None, record: Mapping[str, Any] = None) -> Iterable:
+        """
+        Endpoint does not provide query filtering params, but they provide us
+        updated_at field in most cases, so we used that as incremental filtering
+        during the slicing.
+        """
+
+        if not stream_state or record[self.cursor_field] >= stream_state.get(self.cursor_field):
+            yield record
+
+    def parse_response(
+        self,
+        response: requests.Response,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping]:
+
+        data = self.get_data(response)
+
+        for record in data:
+            updated_at = record.get(self.cursor_field)
+
+            if updated_at:
+                record[self.cursor_field] = datetime.fromtimestamp(
+                    record[self.cursor_field]
+                ).isoformat()  # convert timestamp to datetime string
+
+            yield from self.filter_by_state(stream_state=stream_state, record=record)
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, any]:
+        """
+        This method is called once for each record returned from the API to
+        compare the cursor field value in that record with the current state
+        we then return an updated state object. If this is the first time we
+        run a sync or no state was passed, current_stream_state will be None.
+        """
+
+        current_stream_state = current_stream_state or {}
+
+        current_stream_state_date = current_stream_state.get(self.cursor_field, self.start_date)
+        latest_record_date = latest_record.get(self.cursor_field, self.start_date)
+
+        return {self.cursor_field: max(current_stream_state_date, latest_record_date)}
+
+
+class StreamMixin:
```

Comment:
small comment about naming: can we call it `ChildStreamMixin`, and `slicing_stream` should be `parent_stream_class`?

Reply:
Done